internal/transport/http2_server.go

```diff
@@ -133,6 +133,24 @@ type http2Server struct {
 // underlying conn gets closed before the client preface could be read, it
 // returns a nil transport and a nil error.
 func NewServerTransport(conn net.Conn, config *ServerConfig) (_ ServerTransport, err error) {
+	var authInfo credentials.AuthInfo
+	rawConn := conn
+	if config.Credentials != nil {
+		var err error
+		conn, authInfo, err = config.Credentials.ServerHandshake(rawConn)
+		if err != nil {
+			// ErrConnDispatched means that the connection was dispatched away from
+			// gRPC; those connections should be left open.
+			if err != credentials.ErrConnDispatched {
+				rawConn.Close()
+			}
+			if err == credentials.ErrConnDispatched {
+				return nil, err
+			}
+			rawConn.SetDeadline(time.Time{})
```

Comment:
I spent a lot of brain power trying to match all the logic 1:1 with master to the tee. https://github.com/grpc/grpc-go/blob/master/server.go#L859, here you can see the deadline gets cleared only in the error case. Right now, what I have is the same, and it doesn't matter about the deadline because the conn will be closed. However, I moved the clearing of the deadline to the line after newHTTP2Transport().

Reply:
FWIW: here it also cleared the deadline before your change: https://github.com/grpc/grpc-go/blob/52cea2453436fbb4b962d3cb2da34da7ef6f10c7/server.go#L870

So this behavior was changed before you re-added this.  Having a deadline on a conn means _everything_ on the conn will fail once the deadline elapses, so without this clearing, we end up with connections that only work for several seconds.  Our tests aren't long enough to cover that, which is why they didn't fail.